presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java

```diff
@@ -56,17 +99,72 @@ public void testClusterPools()
             }
 
-            while (reservedPool.getNodes() == 0) {
+            ClusterMemoryPool generalPool = memoryManager.getPools().get(GENERAL_POOL);
+            assertNotNull(generalPool);
+
+            // Wait for the queries to start running and get assigned to the expected pools
+            while (generalPool.getQueries() != 1 || reservedPool.getQueries() != 1 || generalPool.getBlockedNodes() != 2 || reservedPool.getBlockedNodes() != 2) {
                 MILLISECONDS.sleep(100);
             }
 
-            assertTrue(reservedPool.getNodes() > 0);
-            assertTrue(reservedPool.getBlockedNodes() == 0);
-            assertTrue(reservedPool.getTotalDistributedBytes() > 0);
-            assertTrue(reservedPool.getFreeDistributedBytes() > 0);
+            // Make sure the queries are blocked
+            List<QueryInfo> previousQueryInfos = queryRunner.getCoordinator().getQueryManager().getAllQueryInfo();
+            List<QueryInfo> currentQueryInfos = previousQueryInfos;
+            assertEquals(previousQueryInfos.size(), 2);
+            // Check that the pool information propagated to the query objects
+            assertNotEquals(previousQueryInfos.get(0).getMemoryPool(), previousQueryInfos.get(1).getMemoryPool());
+
+            while (!allQueriesBlocked(previousQueryInfos, currentQueryInfos)) {
+                // This is intentionally long to give the queries an opportunity to make progress
+                SECONDS.sleep(5);
+                previousQueryInfos = currentQueryInfos;
+                currentQueryInfos = queryRunner.getCoordinator().getQueryManager().getAllQueryInfo();
+                for (QueryInfo info : currentQueryInfos) {
+                    assertFalse(info.getState().isDone());
+                }
+            }
+
+            // Release the memory in the reserved pool
+            for (TestingPrestoServer server : queryRunner.getServers()) {
+                MemoryPool reserved = server.getLocalMemoryManager().getPool(RESERVED_POOL);
+                // Free up the entire pool
+                reserved.free(reserved.getMaxBytes());
+                assertTrue(reserved.getFreeBytes() > 0);
+            }
+
+            // Make sure both queries finish now that there's memory free in the reserved pool.
+            // This also checks that the query in the general pool is successfully moved to the reserved pool.
+            for (Future<?> query : queryFutures) {
+                query.get();
+            }
+
+            List<QueryInfo> queryInfos = queryRunner.getCoordinator().getQueryManager().getAllQueryInfo();
+            for (QueryInfo info : queryInfos) {
+                assertEquals(info.getState(), FINISHED);
+            }
+
+            // Make sure we didn't leak any memory on the workers
+            for (TestingPrestoServer worker : queryRunner.getServers()) {
+                MemoryPool reserved = worker.getLocalMemoryManager().getPool(RESERVED_POOL);
```

Comment:
That doesn't work since the handling for the general pool and the reserved pool are different